src/bulk/common.ts

```diff
@@ -588,38 +546,61 @@ function executeCommands(
           ? new DeleteOperation(bulkOperation.s.namespace, batch.operations, finalOptions)
           : null;
 
-    if (operation != null) {
-      executeOperation(bulkOperation.s.collection.client, operation).then(
-        result => resultHandler(undefined, result),
-        error => resultHandler(error)
-      );
+    if (operation == null) throw new MongoRuntimeError(`Unknown batchType: ${batch.batchType}`);
+
+    let thrownError = null;
+    let result;
+    try {
+      result = await executeOperation(bulkOperation.s.collection.client, operation);
+    } catch (error) {
+      thrownError = error;
+    }
+
+    if (thrownError != null) {
+      if (!(thrownError instanceof MongoWriteConcernError)) {
+        // Error is a driver related error not a bulk op error, return early
+        throw new MongoBulkWriteError(
+          thrownError,
+          new BulkWriteResult(bulkOperation.s.bulkResult, bulkOperation.isOrdered)
+        );
+      }
+
+      if (thrownError instanceof MongoWriteConcernError) {
+        handleMongoWriteConcernError(
+          batch,
+          bulkOperation.s.bulkResult,
+          bulkOperation.isOrdered,
+          thrownError
+        );
+      }
     }
-  } catch (err) {
-    // Force top level error
-    err.ok = 0;
-    // Merge top level error and return
-    mergeBatchResults(batch, bulkOperation.s.bulkResult, err, undefined);
-    callback();
+
+    mergeBatchResults(batch, bulkOperation.s.bulkResult, thrownError, result);
+    const writeResult = new BulkWriteResult(bulkOperation.s.bulkResult, bulkOperation.isOrdered);
+    bulkOperation.handleWriteError(writeResult);
   }
+
+  bulkOperation.s.batches.length = 0;
```

Comment:
Are you assiging length to 0 to perserve the behavior where `batches.length = 0` after a bulk write?

If so, this is broken when we encounter errors.  Previously, the batchs' length would be accurate.  Now its always equal to the total number of batches.

Reply:
Okay, if this isn't public, I don't understand why we need to clear out the batches after successfully executing all of them.  Can you elaborate?

Reply:
In the unordered case, [we jump over handling errors until the batches are empty](https://github.com/mongodb/node-mongodb-native/pull/4252#discussion_r1781538410).